kyo-core/shared/src/main/scala/kyo/IO.scala

```diff
@@ -56,13 +56,31 @@ object IO:
       *   Implicit frame for the computation.
       * @tparam A
       *   The result type of the main computation.
+      */
+    inline def ensure[A, S](inline f: => Any < IO)(v: A < S)(using inline frame: Frame): A < (IO & S) =
+        ensure(_ => f)(v)
+
+    /** Ensures that a finalizer is run after the computation, regardless of success or failure.
+      *
+      * This version provides the finalizer with information about whether the computation completed successfully or failed with an
+      * exception. The finalizer receives a `Maybe[Throwable]` which will be `Absent` if the computation succeeded, or `Present` if it
+      * failed.
+      *
+      * @param f
+      *   The finalizer function that receives information about potential errors and performs cleanup actions.
+      * @param v
+      *   The computation.
+      * @param frame
+      *   Implicit frame for the computation.
+      * @tparam A
+      *   The result type of the computation.
       * @tparam S
       *   Additional effects in the computation.
       * @return
-      *   The result of the main computation, with the finalizer guaranteed to run.
+      *   The result of the computation, with the finalizer guaranteed to run.
       */
-    def ensure[A, S](f: => Any < IO)(v: A < S)(using frame: Frame): A < (IO & S) =
-        Unsafe(Safepoint.ensure(IO.Unsafe.evalOrThrow(f.unit))(v))
+    inline def ensure[A, S](inline f: Maybe[Throwable] => Any < IO)(v: A < S)(using inline frame: Frame): A < (IO & S) =
```

Comment:
I'm not sure `ensure` wasn't `inline` since that's typically the case for effect suspensions. Note that both `IO.ensure` and `Resource.ensure` now have an overriden version that takes a `Maybe[Throwable]`.